x-pack/plugins/monitoring/server/lib/alerts/fetch_memory_usage_node_stats.ts

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { get } from 'lodash';
+import { AlertCluster, AlertMemoryUsageNodeStats } from '../../alerts/types';
+
+export async function fetchMemoryUsageNodeStats(
+  callCluster: any,
+  clusters: AlertCluster[],
+  index: string,
+  startMs: number,
+  endMs: number,
+  size: number
+): Promise<AlertMemoryUsageNodeStats[]> {
+  const clustersIds = clusters.map((cluster) => cluster.clusterUuid);
+  const params = {
+    index,
+    filterPath: ['aggregations'],
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            {
+              terms: {
+                cluster_uuid: clustersIds,
+              },
+            },
+            {
+              term: {
+                type: 'node_stats',
+              },
+            },
+            {
+              range: {
+                timestamp: {
+                  format: 'epoch_millis',
+                  gte: startMs,
+                  lte: endMs,
+                },
+              },
+            },
+          ],
+        },
+      },
+      aggs: {
+        clusters: {
+          terms: {
+            field: 'cluster_uuid',
+            size,
+          },
+          aggs: {
+            nodes: {
+              terms: {
+                field: 'source_node.uuid',
+                size,
+              },
+              aggs: {
+                index: {
+                  terms: {
+                    field: '_index',
+                    size: 1,
+                  },
+                },
+                avg_heap: {
+                  avg: {
+                    field: 'node_stats.jvm.mem.heap_used_percent',
+                  },
+                },
+                cluster_uuid: {
+                  terms: {
+                    field: 'cluster_uuid',
+                  },
+                },
+                name: {
```

Comment:
I don't think it would be more performant just because we'd have to involve sorting. I think it's pretty performant right now (at least for me), plus we should probably defer any additional optimizations in separate PRs (if the evidence justifies)

Reply:
Fair enough, thanks!